spec/mem/tree_borrows/tree.md

```diff
@@ -0,0 +1,140 @@
+
+```rust
+struct Node {
+    children: List<Node>,
+    /// State for each location
+    location_states: List<LocationState>,
+}
+```
+
+During each memory access, we update states according to the state machine defined in [state_machine.md](state_machine.md). When a node is accessed, each node in the tree can be divided into two disjoint sets: *Child* and *Foreign*, based on its relative position to the accessed node. The Child set includes the node itself and all its descendants, while the Foreign set contains all other nodes.
+
+```rust
+enum NodeRelation {
+    Child, 
+    Foreign,
+}
+```
+
+The state transition depends on both the node relation and the type of access operation.
+
+```rust
+enum AccessKind {
+    Read, 
+    Write,
+}
+```
+
+Then we implement how to actually update and check states in a Tree during each memory access.
+
+```rust
+impl Node {
+    /// Perform state transition on all locations of `self`.
+    /// The `node_relation` indicates whether the accessed node is a child or foreign
+    /// *from the perspective of the `self`*.
+    fn transition(
+        &mut self, 
+        node_relation: NodeRelation,
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size
+    ) -> Result {
+        let offset_start = offset_in_alloc.bytes();
+        for offset in offset_start..offset_start + size.bytes() {
+            let curr_state = self.location_states[offset];
+            let next_state = curr_state.transition(access_kind, node_relation)?;
+            self.location_states.set(offset, next_state);
+        }
+
+        ret(())
+    }
+
+    /// Recusively do state transition on the `self` node and all its descendants.
+    /// The `path` is the path from the `self` to the accessed node.
+    /// The `is_ancestor` indicates whether the `self` is an ancestor of the accessed node.
+    ///
+    /// This method will throw UBs, representing the undefined behavior captured by Tree Borrows.
+    fn access(
+        &mut self,
+        path: Path, // root -> child
+        access_kind: AccessKind,
+        offset_in_alloc: Size,
+        size: Size,
+        is_ancestor: bool,
+    ) -> Result {
+        // Indicates whether the accessed node is a child or foreign
+        // *from the perspective of the `self`*.
+        //
+        // If the `self` is an ancestor of the accessed node, the accessed node is a child of the `self`.
+        // Otherwise, the accessed node is a `foreign` of the `self`
+        let node_relation = if is_ancestor { NodeRelation::Child } else { NodeRelation::Foreign };
+
+        // perform state transition on the `self`
+        self.transition(node_relation, access_kind, offset_in_alloc, size)?;
+
+        let children_num = self.children.len();
+
+        // If the `self` is leaf, stop here.
+        if children_num == Int::ZERO { return ret(()); }
+
+        // The path from the sub-root
```

Comment:
Here things become somewhat more complicated than I expected.

What about the pseudo-code?
* for all children (so no `if children_num == Int::ZERO` needed, the loop will just never run then)
  * if the path starts with this child, do a child access with the path shortened by the first element
  * otherwise, do a child access with the path being `None` (i.e., child is not an ancestor of accessed node)